ledger/internal/evalprefetcher.go

```diff
@@ -0,0 +1,520 @@
+	resources []loadedResourcesEntry
+
+	// err indicates whether any of the balances in this structure have failed to load. In case of an error, at least
+	// one of the entries in the balances would be uninitialized.
+	err error
+}
+
+// accountPrefetcher used to prefetch accounts balances and resources before the evaluator is being called.
+type accountPrefetcher struct {
+	ledger          LedgerForEvaluator
+	rnd             basics.Round
+	groups          [][]transactions.SignedTxnWithAD
+	feeSinkAddr     basics.Address
+	consensusParams config.ConsensusParams
+	outChan         chan loadedTransactionGroup
+}
+
+// loadAccounts loads the account data for the provided transaction group list. It also loads the feeSink account and add it to the first returned transaction group.
+// The order of the transaction groups returned by the channel is identical to the one in the input array.
+func prefetchAccounts(ctx context.Context, l LedgerForEvaluator, rnd basics.Round, groups [][]transactions.SignedTxnWithAD, feeSinkAddr basics.Address, consensusParams config.ConsensusParams) <-chan loadedTransactionGroup {
+	prefetcher := &accountPrefetcher{
+		ledger:          l,
+		rnd:             rnd,
+		groups:          groups,
+		feeSinkAddr:     feeSinkAddr,
+		consensusParams: consensusParams,
+		outChan:         make(chan loadedTransactionGroup, len(groups)),
+	}
+
+	go prefetcher.prefetch(ctx)
+	return prefetcher.outChan
+}
+
+// groupTask helps to organize the account loading for each transaction group.
+type groupTask struct {
+	groupTaskIndex int
+	// balances contains the loaded balances each transaction group have
+	balances []loadedAccountDataEntry
+	// balancesCount is the number of balances that nees to be loaded per transaction group
+	balancesCount int
+	// resources contains the loaded resources each of the transaction groups have
+	resources []loadedResourcesEntry
+	// resourcesCount is the number of resources that nees to be loaded per transaction group
+	resourcesCount int
+	// incompleteCount is the number of resources+balances still pending and need to be loaded
+	incompleteCount int64
+	// err indicates the error code ( if any ) that was received from the ledger during execution
+	err error
+}
+
+// preloaderTask manage the loading of a single element, whether it's a resource or an account address.
+type preloaderTask struct {
+	// account address to fetch
+	address *basics.Address
+	// resource id
+	creatableIndex basics.CreatableIndex
+	// resource type
+	creatableType basics.CreatableType
+	// a list of transaction group tasks that depends on this address
+	groups []*groupTask
+	// a list of indices into the groupTask.balances or groupTask.resources where the address would be stored
+	groupIndices []int
+}
+
+// preloaderTaskQueue is a dynamic linked list of enqueued entries, optimized for non-syncronized insertion and
+// syncronized extraction
+type preloaderTaskQueue struct {
```

Comment:
how's it better than a regular channel? (that would have no additional code/complexity)

Reply:
it's much faster than regular channel - 
in this case, when we write to the queue, we don't need to have *any* types of locks. This is not the case with channel.
Separately, a channel need to be allocated with a large amount of storage, which this queue doesn't need.

Reply:
fair